### PR TITLE
Docs: adds a note about `issubclass` usage in `common_issues.rst`

### PR DESCRIPTION
**Pull request by sobolevn**
It was really not clear for some users that you can use `issubclass` as well.

### CODE REVIEW

File: docs/source/common_issues.rst [hunk 1]
```diff
@@ -402,7 +402,23 @@ style anyway).  We can write the above code without a cast by using
...
+.. code-block:: python
+
+   class MyCalcMeta(type):
+       @classmethod
+       def calc(cls) -> int:
+           ...
+
+   def f(o: object) -> None:
+       t = type(o)  # we must use a variable here
+       if issubtype(t, MyCalcMeta):  # `issubtype(type(o), MyCalcMeta)` won't work
```
Comment (terencehonles) [inline docs/source/common_issues.rst:417]: Is this a typo?

```suggestion
       if issubclass(t, MyCalcMeta):  # `issubclass(type(o), MyCalcMeta)` won't work
```

**Comment by sobolevn** [comment]
@terencehonles oh! Thanks! I make this typo all the time! 😆

**Review by hauntsaninja — COMMENTED** [review]
Thanks for all your work to improve mypy docs. I'm not sure this is the right section: I don't think issubclass is a common solution and I'd like to try and keep the common issues page focussed.

Also, looking more carefully, I believe mypy does now narrow for `type(o) is int`, so this whole section feels out of date...

Anyway, I think what I'd want on the "Common issues" page is a description of how to use `assert` and `cast` to convince mypy that something is of a given type.

Maybe we can add a new top level section on narrowing with detailed examples like in this PR (that we could then link to from common issues). I remember us struggling to find a home for the excellent TypeGuard documentation you wrote, a narrowing section could be a good place for that too.

**Comment by sobolevn** [comment]
> Anyway, I think what I'd want on the "Common issues" page is a description of how to use assert and cast to convince mypy that something is of a given type.

Sure! I will change this PR to add `cast` and `assert`.

> Also, looking more carefully, I believe mypy does now narrow for type(o) is int, so this whole section feels out of date...

Yes, it works for `type(o) is ...` case. But, it does not work for the case in my example:

```python
def f(o: object) -> None:
    if issubclass(type(o), MyCalcMeta): 
        reveal_type(type(o))  # Revealed type is "Type[builtins.object]"
```

Should I open a new issue to track this?

> Maybe we can add a new top level section on narrowing with detailed examples like in this PR (that we could then link to from common issues)

Sounds good to me! I will create one more PR later.

**Comment by sobolevn** [comment]
Moreover, current `mypy` docs [here](https://mypy.readthedocs.io/en/stable/common_issues.html#complex-type-tests) are outdated:

1. This statement:

> Mypy can’t infer the type of o after the type() check because it only knows about isinstance() (and the latter is better style anyway)

This is not true anymore.

2.

```python
 def f(o: object) -> None:
       if type(o) is int:
           o = cast(int, o)
           g(o + 1)    # This would be an error without the cast
           ...
       else:
           ...
```

In this example `cast` is no longer required. The revealed type of `o` is `Revealed type is "builtins.int"`.

I am going to change this example completely. I will find some really useful usecases for `cast()` that cannot be modeled with `isinstance`, `issubclass`, or with custom typeguards.

File: docs/source/common_issues.rst [hunk 2]
```diff
@@ -390,19 +409,11 @@ explicit type cast:
...
     ``o`` was ``Any``.
 
-Mypy can't infer the type of ``o`` after the :py:class:`type() <type>` check
-because it only knows about :py:func:`isinstance` (and the latter is better
-style anyway).  We can write the above code without a cast by using
-:py:func:`isinstance`:
+.. note::
 
-.. code-block:: python
+   You can read more about type narrowing techniques here.
```
Comment (sobolevn) [inline docs/source/common_issues.rst:414]: I will add a link to the new page here in the next PR

File: docs/source/common_issues.rst [hunk 3]
```diff
@@ -367,19 +367,38 @@ above example:
 Complex type tests
 ------------------
 
-Mypy can usually infer the types correctly when using :py:func:`isinstance <isinstance>`
-type tests, but for other kinds of checks you may need to add an
+Mypy can usually infer the types correctly when using :py:func:`isinstance <isinstance>`,
+:py:func:`issubclass <issubclass>`,
+or ``type(obj) is some_class`` type tests,
+and even user-defined type guards,
```
Comment (sobolevn) [inline docs/source/common_issues.rst:373]: I will add a link here as well in the next PR

**Review by hauntsaninja — APPROVED** [review]
Awesome, this looks great!

**Comment by sobolevn** [comment]
@hauntsaninja thanks a lot for your help! 👍 

I will return with the next PR somewhere this week.

**Pull request merged**